pydra/utils/typing.py

```diff
@@ -345,7 +368,26 @@ def coerce_obj(obj, type_):
                     f"Cannot coerce {obj!r} into {type_}{msg}{self.label_str}"
                 ) from e
 
-        return expand_and_coerce(object_, self.pattern)
+        try:
+            return expand_and_coerce(object_, self.pattern)
+        except TypeError as e:
+            # Special handling for MultiInputObjects (which are annoying)
+            if isinstance(self.pattern, tuple) and self.pattern[0] == MultiInputObj:
+                # Attempt to coerce the object into arg type of the MultiInputObj first,
+                # and if that fails, try to coerce it into a list of the arg type
+                inner_type_parser = copy(self)
+                inner_type_parser.pattern = self.pattern[1][0]
+                try:
+                    return [inner_type_parser.coerce(object_)]
+                except TypeError:
+                    add_exc_note(
+                        e,
+                        "Also failed to coerce to the arg-type of the MultiInputObj "
+                        f"({self.pattern[1][0]})",
+                    )
+                    raise e
+            else:
+                raise e
```

Comment:
Another re-ordering of branching.

```suggestion
        except TypeError as e:
            # Special handling for MultiInputObjects (which are annoying)
            if not isinstance(self.pattern, tuple) or self.pattern[0] != MultiInputObj:
                raise e
            # Attempt to coerce the object into arg type of the MultiInputObj first,
            # and if that fails, try to coerce it into a list of the arg type
            inner_type_parser = copy(self)
            inner_type_parser.pattern = self.pattern[1][0]
            try:
                return [inner_type_parser.coerce(object_)]
            except TypeError:
                add_exc_note(
                    e,
                    "Also failed to coerce to the arg-type of the MultiInputObj "
                    f"({self.pattern[1][0]})",
                )
                raise e
```